File: Tests/ConnectorControlStateTests/ReviewModelTests.swift
import XCTest
import ConnectorControlCore
import ConnectorControlTestSupport
@testable import ConnectorControlState

/// Mirror: windows/tests/ConnectorControl.Core.Tests/State/ReviewModelTests.cs.
/// The Review & Apply sheet over a real subscription: a document on disk, a change to it, and the
/// one button that lets the change reach Claude.
@MainActor
final class ReviewModelTests: XCTestCase {

    /// Subscribes to the sample, then publishes a version of it with github gone and dbt's
    /// arguments changed, and reads it into a pending update. Every read here goes through
    /// `recomputePending`, the source watcher's own read, rather than waiting on the watcher:
    /// AppStateCollectionsTests proves the watcher delivers the change.
    private func pending(_ h: AppStateHarness, _ state: AppState) throws -> URL {
        let url = try h.subscribe(state, to: CollectionDocumentSamples.dataTeam)
        var doc = CollectionDocumentSamples.dataTeam
        doc.connectors["github"] = nil
        doc.connectors["dbt"]?.launcher = .local(.init(command: "npx", args: ["-y", "@dbt/mcp@2"], platform: .mac))
        try h.writeDocument(doc, at: url)
        state.recomputePending()
        XCTAssertNotNil(state.pendingUpdates["Data team"])
        return url
    }

    func testRowsDescribeTheDiffAndApplyLandsIt() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        _ = try pending(h, state)

        let model = ReviewModel(state: state, collection: "Data team")
        XCTAssertEqual(model.title, "Update to Data team")
        XCTAssertEqual(model.summary, "removes github; changes dbt")
        XCTAssertEqual(model.rows.map(\.name), ["github", "dbt"], "added, then removed, then changed")
        XCTAssertEqual(model.rows.map(\.kind), [.removed, .changed])
        XCTAssertEqual(model.rows.map(\.id), ["github", "dbt"])

        let github = try XCTUnwrap(model.rows.first)
        XCTAssertEqual(github.before, state.store.collections["Data team"]?.mcps["github"]?.config.editorText())
        XCTAssertNil(github.after)
        let dbt = try XCTUnwrap(model.rows.last)
        XCTAssertEqual(dbt.before, state.store.collections["Data team"]?.mcps["dbt"]?.config.editorText())
        XCTAssertEqual(dbt.after, state.pendingDocument(for: "Data team")?.connectors["dbt"]?.config.editorText())
        XCTAssertEqual(dbt.after?.contains("@dbt/mcp@2"), true)

        XCTAssertNil(model.apply())
        XCTAssertTrue(state.pendingUpdates.isEmpty)
        XCTAssertTrue(model.rows.isEmpty)
        XCTAssertEqual(model.summary, "")
        XCTAssertNil(state.store.collections["Data team"]?.mcps["github"])
        XCTAssertNil(model.apply(), "a second Apply has nothing left to do and nothing to report")
    }

    func testAnAddedConnectorHasNoBeforeSide() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = try h.subscribe(state, to: CollectionDocumentSamples.dataTeam)
        var doc = CollectionDocumentSamples.dataTeam
        doc.connectors["jira"] = .init(launcher: .remote(.init(url: "https://mcp.jira.example/", auth: .automatic,
                                                               package: "mcp-remote", extraArgs: [])))
        try h.writeDocument(doc, at: url)
        state.recomputePending()
        XCTAssertNotNil(state.pendingUpdates["Data team"])

        let model = ReviewModel(state: state, collection: "Data team")
        XCTAssertEqual(model.rows.map(\.name), ["jira"])
        XCTAssertEqual(model.rows.first?.kind, .added)
        XCTAssertNil(model.rows.first?.before)
        XCTAssertEqual(model.rows.first?.after, state.pendingDocument(for: "Data team")?.connectors["jira"]?.config.editorText())
        XCTAssertNil(model.apply())
        XCTAssertEqual(state.store.collections["Data team"]?.mcps["jira"]?.enabled, false, "an added connector arrives off")
    }

    func testApplyRefusesWhenTheSourceMovedUnderTheSheet() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        let url = try pending(h, state)
        let model = ReviewModel(state: state, collection: "Data team")
        XCTAssertFalse(model.sourceMoved)
        XCTAssertEqual(model.rows.map(\.name), ["github", "dbt"])

        // The author commits again while the sheet is open.
        var doc = CollectionDocumentSamples.dataTeam
        doc.connectors["github"] = nil
        doc.connectors["notion"] = nil
        doc.connectors["dbt"]?.launcher = .local(.init(command: "npx", args: ["-y", "@dbt/mcp@3"], platform: .mac))
        try h.writeDocument(doc, at: url)
        state.recomputePending()
        XCTAssertEqual(state.pendingUpdates["Data team"]?.removed, ["github", "notion"])

        XCTAssertEqual(model.apply(), ReviewModel.sourceMovedMessage, "the rows on screen are not what would land")
        XCTAssertTrue(model.sourceMoved)
        XCTAssertNotNil(state.pendingUpdates["Data team"], "nothing was applied")
        XCTAssertNotNil(state.store.collections["Data team"]?.mcps["notion"])

        model.refresh()
        XCTAssertFalse(model.sourceMoved)
        XCTAssertEqual(model.rows.map(\.name), ["github", "notion", "dbt"])
        XCTAssertNil(model.apply())
        XCTAssertTrue(state.pendingUpdates.isEmpty)
        XCTAssertEqual(state.store.collections["Data team"]?.mcps["dbt"]?.config.value(at: JSONPointer(["args", "1"])),
                       .string("@dbt/mcp@3"))
    }

    func testACollectionWithNothingPendingHasNoRows() throws {
        let (h, state) = AppStateHarness.started()
        defer { h.dispose() }
        try h.subscribe(state, to: CollectionDocumentSamples.dataTeam)

        let model = ReviewModel(state: state, collection: "Data team")
        XCTAssertTrue(model.rows.isEmpty)
        XCTAssertEqual(model.summary, "")
        XCTAssertNil(model.apply())
    }
    func testTheKindsGroupTheListInTheOrderTheSummaryReads() {
        XCTAssertEqual(ReviewModel.kinds, [.added, .removed, .changed])
        XCTAssertEqual(ReviewModel.kinds.map(ReviewModel.kindLabel), ["Added", "Removed", "Changed"])
        XCTAssertEqual(ReviewModel.kindLabel(.added), ReviewModel.addedLabel)
        XCTAssertEqual(ReviewModel.kindLabel(.removed), ReviewModel.removedLabel)
        XCTAssertEqual(ReviewModel.kindLabel(.changed), ReviewModel.changedLabel)
    }
    func testTheSheetOwnsItsFooterButtons() {
        XCTAssertEqual(ReviewModel.cancelButton, "Cancel")
        XCTAssertEqual(ReviewModel.refreshButton, "Refresh")
        XCTAssertEqual(ReviewModel.applyButton, "Apply")
    }
}

File: windows/tests/ConnectorControl.Core.Tests/State/ReviewModelTests.cs
using ConnectorControl.Core.State;
using ConnectorControl.Core.Tests.TestSupport;

namespace ConnectorControl.Core.Tests.State;

/// <summary>
/// Mirror: Tests/ConnectorControlStateTests/ReviewModelTests.swift.
/// The Review &amp; Apply sheet over a real subscription: a document on disk, a change to it, and
/// the one button that lets the change reach Claude.
/// </summary>
public class ReviewModelTests
{
    private static readonly TimeSpan Wait = TimeSpan.FromSeconds(8);
    private static readonly TimeSpan WatcherSettle = TimeSpan.FromMilliseconds(300);

    private static void WriteDocument(CollectionDocument doc, string path)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(path)!);
        File.WriteAllBytes(path, doc.Serialize());
    }

    /// <summary>Subscribes to the sample, then publishes a version of it with github gone and
    /// dbt's arguments changed, and waits for that to become a pending update.</summary>
    private static void Pending(AppStateHarness h, AppState state)
    {
        var path = h.Dir.File("data-team.json");
        WriteDocument(CollectionDocumentSamples.DataTeam, path);
        Assert.Null(state.Subscribe(path, null));
        Thread.Sleep(WatcherSettle);
        var sample = CollectionDocumentSamples.DataTeam;
        var connectors = new Dictionary<string, CollectionDocument.Connector>(sample.Connectors, StringComparer.Ordinal);
        connectors.Remove("github");
        var dbt = connectors["dbt"];
        connectors["dbt"] = new CollectionDocument.Connector(
            new CollectionDocument.Launcher.Local("npx", ["-y", "@dbt/mcp@2"], CollectionPlatform.Mac),
            dbt.Env, dbt.Needs, dbt.Additional);
        WriteDocument(new CollectionDocument(sample.Name, sample.Author, sample.Origin, sample.Exported, connectors), path);
        TempDir.BumpModificationTime(path);
        Assert.True(h.Ui.PumpUntil(() => state.PendingUpdates.ContainsKey("Data team"), Wait));
    }

    [Fact]
    public void RowsDescribeTheDiffAndApplyLandsIt()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Pending(h, state);

        var model = new ReviewModel(state, "Data team");
        Assert.Equal("Update to Data team", model.SheetTitle);
        Assert.Equal("removes github; changes dbt", model.Summary);
        // Added, then removed, then changed.
        Assert.Equal(["github", "dbt"], model.Rows.Select(r => r.Name));
        Assert.Equal([ReviewModel.Kind.Removed, ReviewModel.Kind.Changed], model.Rows.Select(r => r.Kind));
        Assert.Equal(["github", "dbt"], model.Rows.Select(r => r.Id));

        var github = model.Rows[0];
        Assert.Equal(state.Store.Collections["Data team"].Mcps["github"].Config.EditorText(), github.Before);
        Assert.Null(github.After);
        var dbt = model.Rows[1];
        Assert.Equal(state.Store.Collections["Data team"].Mcps["dbt"].Config.EditorText(), dbt.Before);
        Assert.Equal(state.PendingDocument("Data team")!.Connectors["dbt"].Config.EditorText(), dbt.After);
        Assert.Contains("@dbt/mcp@2", dbt.After!, StringComparison.Ordinal);

        Assert.Null(model.Apply());
        Assert.Empty(state.PendingUpdates);
        Assert.Empty(model.Rows);
        Assert.Equal(string.Empty, model.Summary);
        Assert.False(state.Store.Collections["Data team"].Mcps.ContainsKey("github"));
        // A second Apply has nothing left to do and nothing to report.
        Assert.Null(model.Apply());
    }

    [Fact]
    public void AnAddedConnectorHasNoBeforeSide()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = h.Dir.File("data-team.json");
        WriteDocument(CollectionDocumentSamples.DataTeam, path);
        Assert.Null(state.Subscribe(path, null));
        Thread.Sleep(WatcherSettle);
        var sample = CollectionDocumentSamples.DataTeam;
        var connectors = new Dictionary<string, CollectionDocument.Connector>(sample.Connectors, StringComparer.Ordinal)
        {
            ["jira"] = new(new CollectionDocument.Launcher.Remote("https://mcp.jira.example/", CollectionDocument.Auth.Auto, "mcp-remote", [])),
        };
        WriteDocument(new CollectionDocument(sample.Name, sample.Author, sample.Origin, sample.Exported, connectors), path);
        TempDir.BumpModificationTime(path);
        Assert.True(h.Ui.PumpUntil(() => state.PendingUpdates.ContainsKey("Data team"), Wait));

        var model = new ReviewModel(state, "Data team");
        Assert.Equal(["jira"], model.Rows.Select(r => r.Name));
        Assert.Equal(ReviewModel.Kind.Added, model.Rows[0].Kind);
        Assert.Null(model.Rows[0].Before);
        Assert.Equal(state.PendingDocument("Data team")!.Connectors["jira"].Config.EditorText(), model.Rows[0].After);
        Assert.Null(model.Apply());
        Assert.False(state.Store.Collections["Data team"].Mcps["jira"].Enabled, "an added connector arrives off");
    }

    [Fact]
    public void ApplyRefusesWhenTheSourceMovedUnderTheSheet()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        Pending(h, state);
        var model = new ReviewModel(state, "Data team");
        Assert.False(model.SourceMoved);
        Assert.Equal(["github", "dbt"], model.Rows.Select(r => r.Name));

        // The author commits again while the sheet is open.
        var sample = CollectionDocumentSamples.DataTeam;
        var connectors = new Dictionary<string, CollectionDocument.Connector>(sample.Connectors, StringComparer.Ordinal);
        connectors.Remove("github");
        connectors.Remove("notion");
        var dbt = connectors["dbt"];
        connectors["dbt"] = new CollectionDocument.Connector(
            new CollectionDocument.Launcher.Local("npx", ["-y", "@dbt/mcp@3"], CollectionPlatform.Mac),
            dbt.Env, dbt.Needs, dbt.Additional);
        var path = h.Dir.File("data-team.json");
        WriteDocument(new CollectionDocument(sample.Name, sample.Author, sample.Origin, sample.Exported, connectors), path);
        TempDir.BumpModificationTime(path);
        Assert.True(h.Ui.PumpUntil(
            () => state.PendingUpdates.TryGetValue("Data team", out var d) && d.Removed.SequenceEqual(["github", "notion"]), Wait));

        // The rows on screen are not what would land.
        Assert.Equal(ReviewModel.SourceMovedMessage, model.Apply());
        Assert.True(model.SourceMoved);
        Assert.True(state.PendingUpdates.ContainsKey("Data team"));   // nothing was applied
        Assert.True(state.Store.Collections["Data team"].Mcps.ContainsKey("notion"));

        model.Refresh();
        Assert.False(model.SourceMoved);
        Assert.Equal(["github", "notion", "dbt"], model.Rows.Select(r => r.Name));
        Assert.Null(model.Apply());
        Assert.Empty(state.PendingUpdates);
        Assert.Equal(JsonValue.String("@dbt/mcp@3"),
            state.Store.Collections["Data team"].Mcps["dbt"].Config.ValueAt(JsonPointer.Parse("/args/1")!));
    }

    [Fact]
    public void ACollectionWithNothingPendingHasNoRows()
    {
        using var h = new AppStateHarness();
        using var state = h.Create();
        var path = h.Dir.File("data-team.json");
        WriteDocument(CollectionDocumentSamples.DataTeam, path);
        Assert.Null(state.Subscribe(path, null));

        var model = new ReviewModel(state, "Data team");
        Assert.Empty(model.Rows);
        Assert.Equal(string.Empty, model.Summary);
        Assert.Null(model.Apply());
    }
    [Fact]
    public void TheKindsGroupTheListInTheOrderTheSummaryReads()
    {
        Assert.Equal([ReviewModel.Kind.Added, ReviewModel.Kind.Removed, ReviewModel.Kind.Changed], ReviewModel.Kinds);
        Assert.Equal(["Added", "Removed", "Changed"], ReviewModel.Kinds.Select(ReviewModel.KindLabel));
        Assert.Equal(ReviewModel.AddedLabel, ReviewModel.KindLabel(ReviewModel.Kind.Added));
        Assert.Equal(ReviewModel.RemovedLabel, ReviewModel.KindLabel(ReviewModel.Kind.Removed));
        Assert.Equal(ReviewModel.ChangedLabel, ReviewModel.KindLabel(ReviewModel.Kind.Changed));
    }
    [Fact]
    public void TheSheetOwnsItsFooterButtons()
    {
        Assert.Equal("Cancel", ReviewModel.CancelButton);
        Assert.Equal("Refresh", ReviewModel.RefreshButton);
        Assert.Equal("Apply", ReviewModel.ApplyButton);
    }
}
